plutus-core-spec/plutus-core-specification.tex

```diff
@@ -1093,50 +1098,253 @@ \subsection{Type erasure}
+result of \textit{discharging} the closure, ie, recursively replacing
+all of the variables occurring in $V$ by their bindings in
+$\rho$. Note that variables are bound to closures which may themselves
+need to be discharged: hence the use of the word ``recursively''. It
+can be shown that when any closure produced by the CEK machine is
+discharged it yields a value.
+
+\subsection{Definition of the CEK machine}
+The CEK machine is described in Figure~\ref{fig:cek-machine}.  It is
+  is very similar to the CK machine of Figure~\ref{fig:typed-ck-machine},
+  except that
+\begin{itemize}
+\item Every rule now involves environments
+\item A new rule has been added to deal with the evaluation of variables
+\item The rule involving lambda application has changed to replace substitution with
+the use of an extended environment
+\item When the machine terminates without error it now
+returns a closure $(V,\rho)$ rather than a value.  The invoker
+of the machine may wish to discharge this closure to obtain a value.
+\end{itemize}
+\noindent The major changes are highlighted in \blue{blue}. To evaluate a program
+\texttt{(program $v$ $M$)} we check the version number $v$ and then evaluate the body $M$
+with an empty stack and environment: $\cdot;[] \compute M$.
+
+\subfile{figures/TypedCekMachine.tex}
+
+\subsubsection{Note. Built-in functions in the CEK machine}
+\label{sec:typed-cek-builtins}
+In the CK machine the arguments of built-in functions are always
+closed values, but in the CEK machine arguments are closures.  We
+assume that the mechanism for evaluating built-in functions in the CEK
+machine can deal with closures, and that
+\begin{itemize}
+\item Evaluation of builtin functions always terminates.
+
+\item For closures $C_1, \ldots, C_n$, evaluation of
+  \texttt{(builtin}\ \textit{bn} $A_1 \ldots A_k\  C_1 \ldots C_n$ \texttt{)}
+fails in the CEK machine if and only if evaluation of
+  \texttt{(builtin}\ \textit{bn} $A_1 \ldots A_k\ \discharge{C_1} \ldots \discharge{C_n}$ \texttt{)}
+fails according to Figure~\ref{fig:builtins}.
+
+\item For closures $C_1, \ldots, C_n$,
+  \texttt{(builtin}\ \textit{bn} $A_1 \ldots A_k\  C_1 \ldots C_n$ \texttt{)}
+evaluates to a closure $C$ in the CEK machine if and only if 
+  \texttt{(builtin}\ \textit{bn} $A_1 \ldots A_k\  \discharge{C_1} \ldots \discharge{C_n}$ \texttt{)}
+evaluates to a value $V$ (according to Figure~\ref{fig:builtins})
+such that $\discharge{C} = V$.
+
+\end{itemize}
+\noindent (One way to achieve this would be to discharge all of the closures and
+then use the same mechanism as in the CK machine, but this would
+reintroduce the very inefficiencies which the CEK machine is designed
+to avoid.)  A future version of this document will contain a
+specification of built-in function evaluation which will cover this
+issue in detail.
+
+\subsubsection{Comparison of the CK and CEK machines}
+If the CK and CEK machines are run with the same input then we can prove
```

Comment:
"same input" here means starting the CEK machine with an empty environment, presumably.

Reply:
I just meant the same program, but I'll clarify that.